### PR TITLE
Replace clients `host` and `scheme` properties to cater to S3-Compatible backends

### PR DESCRIPTION
**Pull request by oleiade**
## What?

This Pull Request is massive, but it does a single thing: replace the existing `AWSClient.host` and `AWSClient.scheme` properties with a single `endpoint` property both on the `AWSClient`, and the `AWSConfig`.

This Pull Request adds an option to `AWSConfig`:
```
const awsConfig = new AWSConfig({
    region: __ENV.AWS_REGION,
    accessKeyId: __ENV.AWS_ACCESS_KEY_ID,
    secretAccessKey: __ENV.AWS_SECRET_ACCESS_KEY,
    sessionToken: __ENV.AWS_SESSION_TOKEN,
    endpoint: 'https://fra1.digitaloceanspaces.com'
})
```

In the ☝🏻 example, every request will now be sent to `https://fra1.digitaloceanspaces.com` instead of `https://amazonaws.com`.

## Why?

This is an attempt to replicate the constructs in the node AWS SDK, as discussed in #57, and expose an `endpoint` property. Using this property, users can now indicate to the client classes instances via the `AWSConfig.endpoint` property or by setting it directly on the client the actual URL endpoint it should use to talk to AWS.

The reason for that is to cater to AWS-compatible services and stacks, such as in #21 #57 #50, and allow for a nicer integration of the SDK with services such as DigitalOcean spaces and Cloudflare R2 offer S3-compatible APIs or with Localstack, for instance.

This feature does **not** mean that we officially support alternative S3-compatible APIs, but it should offer the possibility to whoever needs to use it to do so at their own risk.

### CODE REVIEW

**Comment by oleiade** [comment]
Hi @Schachte, and @w0rd-driven 👋🏻 

Just a heads-up that I believe this PR should help address your recent issues interoperating with Cloudflare R2 and Digital Ocean spaces.

Could you take it out for a spin, and let me know how it goes? 🙏🏻 

Thank you! 🙇🏻

**Comment by Schachte** [comment]
Awesome work! I'll review this guy today. One callout without actually having dug too deep yet, was this comment:

https://github.com/grafana/k6-jslib-aws/issues/50#issuecomment-1590816808

> endpoint definitely shouldn't have the scheme in it. 

I can browse to see if that is problematic as the example above prefixes with `https`. I know the default scheme is assumed https, but wasn't sure if there were weird cases where it would end up like `https://https://` etc.

File: src/internal/config.ts [hunk 1]
```diff
@@ -30,21 +31,13 @@ export class AWSConfig {
...
-
-    // FIXME: Should really be called "host" instead. When used
-    // with localstack we pass a complete host (hostname:port) here.
     /**
      * The AWS hostname to connect to.
      *
      * @type {string} ['amazonaws.com']
      */
-    endpoint: string = 'amazonaws.com'
+    // endpoint: string = 'amazonaws.com'
```
Comment (Schachte) [inline src/internal/config.ts:39]: nit: remove stale comment

File: src/internal/config.ts [hunk 2]
```diff
@@ -185,7 +178,8 @@ export interface AWSConnectionOptions {
      *
      * @type {string}
      */
-    endpoint?: string
+    // endpoint?: string
```
Comment (Schachte) [inline src/internal/config.ts:181]: same nit about comment

File: src/internal/endpoint.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    // Default protocol, this can be globally changed as per application requirements
+    private static readonly DEFAULT_PROTOCOL = 'https' // Set this as per AWS.config
+
+    /**
+    //  * Constructs a new Endpoint instance.
+     *
+     * @param {string} endpoint - The URL to construct an endpoint from. If the URL omits a protocol, the default protocol will be used.
+     */
+    constructor(endpoint: string) {
+        const isDefaultProtocol =
```
Comment (Schachte) [inline src/internal/endpoint.ts:18]: I think this constructor can be simplified a bit if you leverage `URL`.  

```ts
const completeUrl = endpoint.startsWith('http://') || endpoint.startsWith('https://')
    ? endpoint
    : `${Endpoint.DEFAULT_PROTOCOL}://${endpoint}`;

const url = new URL(completeUrl);

this._protocol = url.protocol.slice(0, -1);
this._hostname = url.hostname;
this._port = url.port ? parseInt(url.port) : undefined;
```

**Review by Schachte — APPROVED** [review]
Looks good! Thanks for the fast follow-up. Left just a couple nits.

File: src/internal/endpoint.ts (hunk 3 above)
Comment (oleiade) [reply src/internal/endpoint.ts:18]: Unfortunately `URL` is not yet available in the k6 JS runtime, hence doing it manually instead 👍🏻

Comment (Schachte) [reply src/internal/endpoint.ts:18]: 👍 . Heads up, you might want to remove
https://github.com/grafana/k6-jslib-aws/pull/61/files/6d50785f5a9123ff5c2ffca766fa8b5fe4969256#diff-cb88677cfb485379a627afd78775c5daebe4bf9085f172f84a87549d8b62e31fR97

Comment (oleiade) [reply src/internal/endpoint.ts:18]: Good catch 👀

**Pull request merged**